### PR TITLE
GRPO: put vLLM's truncated-sampling log-probs on the trainer's support (fixes the truncated-support bias in the importance ratio)

### LINKED ISSUE #6789: GRPO: vLLM importance-sampling ratio is biased when top_p/top_k/min_p truncate sampling
When `top_p < 1`, `top_k > 0` or `min_p` is set, GRPO's vLLM importance-sampling correction compares two log-probabilities computed over **different distributions**, so the ratio is biased. vLLM returns log-probs renormalized over the truncated (nucleus) distribution, while the trainer recomputes them over the full vocabulary.

Two other RL frameworks have now shipped a training-side fix for exactly this, and the native vLLM fix is merged but not yet in a release TRL supports.

### Where the mismatch is

Generation side asks vLLM for post-processor log-probs, in both colocate and server mode:

https://github.com/huggingface/trl/blob/main/trl/generation/vllm_generation.py#L365

Training side divides by temperature and takes a full-vocab log-softmax:

https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L1530-L1534

`processed_logprobs` means "after every logits processor", which includes top-p/top-k truncation followed by renormalization over the surviving tokens. Temperature is handled symmetrically (line 1532 mirrors it, and that symmetry was the fix for #4159); truncation is not. For a sampled token that survives the nucleus, vLLM's log-prob is therefore *higher* than the trainer's by `-log(sum of surviving probability mass)`, which is a per-token, systematically negative bias in `log π_θ - log π_behaviour`.

`importance_sampling_ratio` multiplies the per-token loss directly:

https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L3208-L3209

so the effect is a silent, token-dependent down-weighting of the policy gradient rather than a crash.

### Scope

- Defaults are safe: `top_p=1.0`, `top_k=0`, `min_p=None`, so nothing is truncated and the two distributions agree.
- `AsyncGRPOTrainer` is affected more severely. Server launch passes `--logprobs-mode processed_logprobs`, and those generator log-probs are not an optional correction there: they are the denominator of the policy ratio itself, with no flag guarding them.

  ```python
  log_ratio = log_probs - old_log_probs   # async_grpo_trainer.py#L945
  coef_1 = torch.exp(log_ratio)
  ```

  `AsyncGRPOConfig` exposes `top_p`/`top_k`/`min_p` like GRPO, and it logs no `sampling_logp_difference` equivalent, so there is nothing to notice the drift with.
- Affected as soon as a user sets any of them, which `top_p=0.9`/`top_k=50` recipes sometimes do.
- Only active with `vllm_importance_sampling_correction=True`.
- Same `logprobs_mode="processed_logprobs"` is set in `trl/scripts/vllm_serve.py#L361`, so server mode is affected identically.

### It is already observable

The existing metric spikes under truncation, since it measures precisely this gap:

`sampling/sampling_logp_difference/mean` and `.../max` (grpo_trainer.py, `_generate_and_score_completions`)

A quick check: run GRPO+vLLM with `top_p=1.0` and then `top_p=0.8`, and compare that metric. If it jumps with no other change, the bias is real. Worth doing before we pick a fix.

### Prior art

- vLLM [#49577 "Mask Replay"](https://github.com/vllm-project/vllm/pull/49577) is **merged** (2026-08-13) but landed after v0.27.1 (2026-08-11), so it is not in any release covered by TRL's `vllm>=0.18.0,<=0.27.1` pin. We cannot rely on it yet.
- slime shipped a training-side mask in v0.3.1.
- prime-rl [#3235 "feat: top-p/top-k train sampling with sampling replay"](https://github.com/PrimeIntellect-ai/prime-rl/pull/3235) is building the same thing independently.

### Possible directions

1. Apply the same truncation to the training-side logits before the log-softmax, so both sides normalize over the same support. Mirrors what slime and prime-rl do, and keeps the temperature-symmetry precedent from #4159.
2. Ask vLLM for raw log-probs and apply temperature and truncation ourselves on both sides.
3. Wait for Mask Replay and lift the ceiling to vLLM 0.28. Doesn't help anyone on a supported version today.
4. Do nothing to the math, but warn when `vllm_importance_sampling_correction=True` is combined with truncation, so the bias is at least not silent.

I'd like a repro of the metric jump before committing to one.

### Possibly related

- #4159
- #4772

### PR DESCRIPTION
**Pull request by GuoCheng24**
# What does this PR do?

Fixes the truncated-support bias in the vLLM importance-sampling correction.

When sampling is truncated (`top_p < 1`, `top_k > 0`, `min_p > 0`), vLLM's `processed_logprobs` are normalised over the tokens that survived the truncation, while the trainer recomputes log-probs over the full vocabulary. The ratio `exp(logp_trainer - logp_vllm)` then carries a factor equal to the kept probability mass of each position, which is not a policy change. Measured on Qwen2.5-1.5B-Instruct with vLLM 0.28 (2,048 tokens per row, temperature 1.0, trainer side = HF bf16):

| sampling | kept mass | ratio, full-vocab trainer log-probs: mean / median / outside [0.9, 1.1] | ratio, renormalised over the replayed support |
|---|---|---|---|
| top_p=0.8, top_k=1024 | 0.895 | 0.896 / 0.895 / 51.8% | 1.001 / 1.000 / 4.3% |
| top_k=20 | 0.965 | 0.965 / 0.995 / 16.2% | 1.000 / 1.000 / 6.2% |
| top_p=1.0, top_k=1024 (control) | 0.994 | 0.995 / 1.000 / 6.9% | 1.000 / 1.000 / 5.9% |

The full-vocabulary mean equals the kept mass; the residual after renormalisation is the ordinary engine-vs-trainer noise of the control row.

vLLM >= 0.28 can return the kept token ids of every generated token (`return_sampling_mask`, vllm-project/vllm#49577, which the RFC vllm-project/vllm#42259 added for exactly this use). This PR adds `GRPOConfig.vllm_importance_sampling_support`:

- `"vocab"` (default): unchanged behaviour.
- `"sampled"`: colocate vLLM is created with `return_sampling_mask=True`; `VLLMGeneration` extracts the per-token kept ids (`extract_sampling_supports`, exposed as `VLLMGeneration.sampling_supports` so that the 4-tuple return that other trainers unpack positionally is untouched); `GRPOTrainer` pads them to `(B, T, K)` and, in the pass that computes `old_per_token_logps`, also computes `log_kept_mass` — the log of the probability mass vLLM sampled from at each position (new `trl.trainer.utils.log_kept_mass`, `0.0` where a row has no kept token). That value is **added to vLLM's `processed_logprobs`**, lifting them onto the full-vocabulary support before the importance-sampling ratio, the `sampling/*` metrics and the off-policy mask are formed. The trainer's own log-probs — old, current and reference — are not changed, so the policy-gradient ratio `exp(per_token_logps - old_per_token_logps)` is untouched.

Constraints, mirrored from vLLM and validated in `GRPOConfig.__post_init__`: colocate mode only (the OpenAI-compatible server does not return the mask), `top_k > 0` (vLLM bounds the replayed set with it; a large value such as 1024 leaves `top_p` in charge), `temperature > 0`. When the returned supports do not line up with the completions (tool turns, custom rollouts), the trainer logs a warning once and falls back to the full vocabulary rather than normalising over the wrong set.

Relation to #6880: that PR warns when the bias is present; this one removes it where vLLM can replay the support. The two are complementary.

## Tests

- `tests/test_utils.py::TestLogKeptMass`: a full support gives exactly zero; a subset gives the log of its probability mass; `-1` padding is ignored and an empty support gives `0.0`.
- `tests/test_grpo_trainer.py::test_sampling_support_option_is_validated`: the three constraint errors and the valid configuration.
- `tests/test_grpo_trainer.py::test_sampling_support_renormalisation_removes_truncation_term[vocab|sampled]`: drives the real trainer with a stand-in sampler that computes vLLM-style log-probs over a 3-token support; with `"sampled"` the recorded `sampling/sampling_logp_difference/mean` is ~0, with `"vocab"` it equals the mean log kept mass; in both modes `clip_ratio/region_mean` must stay exactly 0 (one optimisation step per generation), which is the assertion that fails on a version that renormalises the trainer's log-probs instead.
- End to end with a real colocated vLLM 0.28 (Qwen2.5-0.5B-Instruct, GSM8K prompts, top_p=0.8, top_k=1024, temperature 1.0, 6 steps): `sampling/sampling_logp_difference/mean` per step is 0.052, 0.053, 0.040, 0.044, 0.037, 0.031 with `"vocab"` and 0.008, 0.009, 0.006, 0.007, 0.006, 0.005 with `"sampled"` (0.043 vs 0.007 on average), the residual being the ordinary engine-vs-trainer noise, and `clip_ratio/region_mean` stays at most 0.0005 in both modes, i.e. the policy-gradient ratio is untouched.

Fixes #6789

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. — #6789, and vllm-project/vllm#42259 for the vLLM side.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@qgallouedec — this is the trainer-side half of the #6789 thread.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2674,6 +2711,7 @@ def _generate_and_score_completions(
                     batch_size,
                     num_images=num_images,
                     num_tiles=num_tiles,
+                    sampling_support_ids=sampling_support_ids,
```
[the remark this answers is not in the document]
Comment (GuoCheng24) [reply trl/trainer/grpo_trainer.py:2716]: Right on both counts, thank you. `old_per_token_logps` is also π_old for `exp(per_token_logps - old_per_token_logps)`, so renormalising it moved the kept-mass factor from the engine ratio into the policy-gradient ratio; and an all-padding support row produced `-inf` in the log-sum-exp.

Fixed by turning the correction around: the trainer's log-probs stay full-vocabulary everywhere (old, current, reference), and the one pass that already has the logits also returns `log_kept_mass` — the log of the probability mass vLLM sampled from — which is added to vLLM's `processed_logprobs`. That lifts the sampler's numbers onto the same support as the trainer's, so the importance-sampling ratio, the `sampling/*` metrics and the off-policy mask all compare like with like, while `old_per_token_logps` is untouched. A row with no kept token gets a correction of `0.0`, not `-inf`.

The trainer test now also asserts `clip_ratio/region_mean == 0` in both modes (one optimisation step per generation, so the policy-gradient ratio must be exactly 1); on the previous commit that assertion fails in the `sampled` case for exactly the reason you describe.

File: trl/trainer/utils.py [hunk 2]
```diff
@@ -498,6 +498,33 @@ def flush_left(mask: torch.Tensor, *tensors: torch.Tensor) -> torch.Tensor | tup
...
+
+    Returns:
+        `torch.Tensor`:
+            `log m` with shape `(...)`, in float32, always `<= 0`.
+    """
+    valid = support_ids >= 0
+    kept = torch.gather(logits, dim=-1, index=support_ids.clamp(min=0)).float()
+    kept = kept.masked_fill(~valid, float("-inf"))
+    log_mass = torch.logsumexp(kept, dim=-1) - torch.logsumexp(logits.float(), dim=-1)
+    return torch.where(valid.any(dim=-1), log_mass, torch.zeros_like(log_mass))
```
[the remark this answers is not in the document]
Comment (GuoCheng24) [reply trl/trainer/utils.py:530]: Agreed — the one-shot `logits.float()` was an fp32 copy of the whole `(B, T, V)` tensor. `log_kept_mass` now takes the full-vocabulary log-sum-exp row by row, exactly as `selective_log_softmax` does next to it, so the fp32 copy is one row wide at a time; fp32/fp64 logits skip the copy entirely. Values are unchanged (83f908f).

[the remark this answers is not in the document]
Comment (GuoCheng24) [reply trl/trainer/utils.py:530]: Measured on an RTX 4090 with a GRPO-shaped input (B=4, T=384, V=151,936, bf16 logits of 0.43 GiB, K=64): peak extra allocation 1.74 GiB with the one-shot version, 0.44 GiB row by row; outputs identical to 1e-4.

**Comment by qgallouedec** [comment]
Thanks.
Before anything else, please fill in the AI disclosure section of the template.

**Comment by qgallouedec** [comment]
Closing. The AI writing disclosure section was removed from the PR description, we ask first-time contributors to fill it in: https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#ai-usage-policy

**Comment by GuoCheng24** [comment]
Sorry for the round trip, and thanks for pointing at the policy.

The disclosure section is filled in now — **AI-assisted** — along with the rest of the template. What happened is that I renamed the PR and replaced the description with `--body-file` in the same step, and the template went with it; I did not realise the section had gone until you closed this. That was careless on my part, not an attempt to skip it.

Nothing about the change itself has moved since your last look (head is still `83f908f`): it adds `GRPOConfig.vllm_importance_sampling_support`, default `"vocab"` so existing behaviour is untouched, and under `"sampled"` it adds the log of vLLM's kept probability mass to vLLM's own log-probs so they sit on the full-vocabulary support before the importance-sampling ratio is formed. The trainer's log-probs are not touched, so the policy-gradient ratio is unchanged — the two Bugbot findings on that point are fixed and answered in the thread with measurements.

I do not have permission to reopen this myself. Could you reopen it, or would you rather I open a fresh PR from the same branch?
